### ISSUE TITLE
Hashtags

### ISSUE THREAD

**Issue by buddh4** [humhub/humhub#3936]
Implement hashtag richtext plugin.

**Comment by nsturtz2** [comment]
I found the code it is at protected/vendor/npm-asset/at.js and there is a hashtags file protected/vendor/npm-asset/at.js/examples/hashtags.html how do I enable it to use in Humhub?

**Comment by MasterMindNET** [comment]
What would you like to tag using a hashtag? there is an `at` (@) sign to reference someone or space.
What would be the main purpose of this plugin (hashtag richtext plugin)?

**Comment by buddh4** [comment]
Similar to facebook hashtags users can use hashtags as #summer #newyear in the richtext which will link to a dashboard like stream with all content also tagged with this hashtag.

**Comment by Carl2213** [comment]
Hashtags are like topics, only universal. Currently, the topics feature is contained within the content container of the stream (the space or user's profile).

Hashtags would function exactly the same way but on a global level. This would help with discoverability and curating stream content. Hashtags are also easier to use. Simply #summer instead of adding a topic to the post (hashtags can also be triggered in comments) or having to manage topics.

I think it would be possible to rework the Topics feature into Hashtags. Remove the content container limit to make them global, then add #logic to the richtext in order to automatically create a search/stream filter with the hashtag when the # character is detected.

Content from the user's profile or joined spaces can be joined at the top. Further filters (friends/spaces) can further customize the stream/search results.

**Comment by Carl2213** [comment]
Further optional extensions:

- calculate the number of mentions of a particular hashtag to feed a #Trending widget with the most used/discussed hashtags
- Like mentioning, auto suggest hashtags in the richtext when the # character is typed, along with the current usage number of each hashtag

**Comment by buddh4** [comment]
Also the topic feature can be used as moderator only tool to tag and categorize content.

**Comment by Carl2213** [comment]
Maybe both can be kept, but the underlying logic can be copied and modified and renamed to hashtags?

**Comment by ArchBlood** [comment]
@buddh4 on the topic of topics feature should also be able to hide topics as well, but another issue should be opened for that :thinking:

**Comment by buddh4** [comment]
@ArchBlood not sure what you mean, but topics should be rather hidden for users who can't create topics unless there are topics available in a space.

For me the differences between hashtags and topics are:

- Hashtag are user generated and not moderated
- Hashtags are added directly in a richtext field and are rendered as simple links
- Hashtags are global (not container related) 
- There probably will be more hashtags than topics
- Topic are rather comparable to the labels here on github

**Comment by ArchBlood** [comment]
> @ArchBlood not sure what you mean

What I mean by hide is, default topics with viewability, meaning being able to set a default topic to be viewable or hidden making the post not viewable by users other than admin and moderator permission leveled users

**Comment by samuk** [comment]
#want #please

**Comment by samuk** [comment]
@luke- If resources become available to impliment ActivityPub https://github.com/humhub/humhub/issues/3150

Perhaps the development of #Tags should be included within that work as much of the Fediverse operates on hashtags.

**Comment by luke-** [comment]
If would prefer to use the existing Topics for this. A new additional categorization would probably be too confusing.

**Comment by samuk** [comment]
Thanks. Topics would never be compatible with ActivityPub though? 

We've never used the existing topic functionality as it did not meet our need to have user-defined categorization.

**Comment by luke-** [comment]
I'm not sure about that, probably we could pass the HumHub Topics in the ActivityPub.

If an external post comes with tags, we could also transfer them to Humhub Topics.

**Comment by ArchBlood** [comment]
Here's an example for the JavaScript for Hashtags, no this hasn't been tested;

```javascript
humhub.module('content', function(module, require, $) {
    var Widget = require('ui.widget').Widget;

    // Function to extract hashtags from text
    function extractHashtags(text) {
        var regex = /#(\w+)/g; // Match hashtags starting with #
        var hashtags = text.match(regex) || []; // Extract hashtags from the text

        return hashtags;
    }

    // Function to handle posts and extract hashtags
    function handlePosts(context) {
        $(context).find('.wall-entry').each(function() {
            var postText = $(this).find('.content').text();
            var hashtags = extractHashtags(postText);

            // Do something with the extracted hashtags (e.g., display them, perform an action)
            console.log('Hashtags found:', hashtags);
        });
    }

    // Initialize the module
    module.export({
        init: function() {
            // Handle posts when the page is loaded and on PJAX success
            $(document).ready(function() {
                handlePosts(document);
            });

            // Handle PJAX success event to re-bind functionality on dynamically loaded content
            $(document).on('humhub:ready', function(event, data) {
                handlePosts(data.$newContent);
            });
        }
    });
});
```

**Comment by ArchBlood** [comment]
Currently implementing something similar into the forums module, currently it works as I needed it to, the only difference is that it works similarly to a mention instead of a hashtag but without the added notification. But would definitely be much easier if there were already an existing extension within HumHub that could mimic this function.

![Screenshot_1](https://github.com/user-attachments/assets/03ad2339-44c1-44e8-b728-d845b5b7f0f1)

**Comment by ArchBlood** [comment]
@luke- I don't believe this is a good idea for a module as it seems to be rather complicated, it works rather well within the core of Humhub using a custom helper class along with the search function within v1.16.0, the only issue is that using `RichText::output()` can not handle this, but using `RichTextToHtmlConverter::process()` enables the use without error, to actually make it work it would require an overwrite injection `getViewPath()` method if implemented within a module which would cause issues if other modules also inject into the `wallEntry.php` post widget view.

![Screenshot_1](https://github.com/user-attachments/assets/2207bbc0-508d-44ca-b9fe-8876964e2e0b)

**Comment by ArchBlood** [comment]
@luke- I pushed the module to the marketplace so this can be closed once it is reviewed.

**Comment by luke-** [comment]
@ArchBlood Hmm, I'm having a bit of difficulty with the module.

- Overwriting the view or core components has a huge potential to cause very understandable issues, especially when it comes to updates.

- With our topics, which should be branded more as hashtags in the future, there is a conflict of usability here. 
We want users to really use the topics as keywords, because these are neatly stored in the database. Topics has also have some extended features like filter a stream.

- I don't want to allow a module that later becomes obsolete due to changes to the core and users lose all their previously defined hashtags.  

Not sure how to proceed here...

**Comment by ArchBlood** [comment]
> @ArchBlood Hmm, I'm having a bit of difficulty with the module.
> 
> * Overwriting the view or core components has a huge potential to cause very understandable issues, especially when it comes to updates.
> * With our topics, which should be branded more as hashtags in the future, there is a conflict of usability here.
>   We want users to really use the topics as keywords, because these are neatly stored in the database. Topics has also have some extended features like filter a stream.
> * I don't want to allow a module that later becomes obsolete due to changes to the core and users lose all their previously defined hashtags.
> 
> Not sure how to proceed here...

I completely agree with all points, the issue though is with topics themselves they are harder to deal with because of visibility and are within their own field completely which is hidden currently, sure in the future this may change slightly but it doesn't give the user their requested traditional hashtags function. With the hashtags module we also create and store the keyword within the database, the only thing that the Hashtags module does not do is filter the stream, it wasn't made for this nor are hashtags meant to, they are kind of hyperlink to search all possible posts that have that hashtag nothing else.

To counteract with some of the points I could do something similar as I have done within the Verified module which I have already done partly when it comes to a helper class that handles creating the hashtag links. Hashtags shouldn't be lost as they're stored within the database under `hashtag` so lose of them shouldn't be a problem. As for the main concern about the topics vs hashtags I'd also like to refer to the same which can be said about having multiple options for the search function, more advanced or not, currently there is the core version, the [Advanced Search](https://marketplace.humhub.com/module/advanced-search/description) and [Apache SOLR](https://marketplace.humhub.com/module/solr/description) wouldn't this also be considered as a conflict of usability in the future?

**Comment by Eladnarlea** [comment]
+1 on the HumHub topics will most likely become hashtags in the future, therefore I am not convinced of this module... although it seems to be a cool feature at the moment, I think it conflicts too much with our future vision of topics.

**Comment by marc-farre** [comment]
Independently of the question about core or module, I can see that no one is using topics.
Even on https://community.humhub.com/dashboard I don't see any topics except on the "official" [Announcements space](https://community.humhub.com/s/announcements/).
And even when topics are added to content, I don't think anybody clicks on it.

So the topic concept is great, but I think there is a problem of UX with it.
And replacing them with hashtags (perhaps keeping the same models, DB and stream filtering logic) might help, because:
- hashtags are common in social networks, so people already know how to use them
- it's more visual to see hashtag links into the text instead of buttons that looks like a labels
- it's quicker and easier to type `#` when typing text than having to think to add topics

However, this will work with posts or content having rich text, but not with other content.
So maybe we could keep topics as they are, but:
- remove permissions to create topics (if you can create content, you can create a topic)
- when writing a word starting with `#`, a topic is added to the content
- remove topics manager (because there might be too many topics and anybody can add topics by writing a hashtag)
- for posts, display topics as hashtags directly into the text

For existent post topics, we could move them to hashtags style with a migration which will add them to the top of the text:

Before:
![image](https://github.com/user-attachments/assets/afc0d686-64ca-4ece-a2ab-9ad417079e7d)

After:
![image](https://github.com/user-attachments/assets/d3529030-bbd4-46a9-ac6c-aa267402450b)

**Comment by samuk** [comment]
As a user I'd love to see some progress on tags/topics.  In there present state I don't find topics useful. The core plans sound good. 

In the short term I would also use a sub-optimal module based version if the core based work is a year or two away.

Or can the core work be added to the workplan/ roadmap more quickly than that?

**Comment by luke-** [comment]
We are currently unable to prioritise this topic.
But we also can't offer users a module that will later lead to them losing all previously used tags.

Perhaps this is still useful for some installations. They could explicitly install it manually. But I wouldn't actively "promote" it via the Marketplace.

**Comment by samuk** [comment]
Is the topic/tag work in core something @marc-farre could work on if I was able to fund it? Or is it something you would only want to do in-house?

**Comment by luke-** [comment]
We need to develop a detailed concept & UI for this. This is an issue for the team and difficult as an individual contribution.

**Comment by ArchBlood** [comment]
> We are currently unable to prioritise this topic. But we also can't offer users a module that will later lead to them losing all previously used tags.
> 
> Perhaps this is still useful for some installations. They could explicitly install it manually. But I wouldn't actively "promote" it via the Marketplace.

If it is feared that the users will lose their previously used tags I don't mind providing a migration once the revised topics is introduced, but you are correct, promoting such a module within the marketplace wouldn't be the best of ideas in this sense so I retract the statement and pushed version from the partners section.

**Comment by CarlSinclair** [comment]
I agree with @marc-farre. While topics may be a powerful feature, that doesn't ultimately matter if it's not used. Rebranding it as hashtags and changing its frontend appearance is the best way forward IMO. That's the easiest and fastest way to solve both problems.

Hashtags need no UI for adding them to a post since they're added directly in the richtext editor. But they need to be highlighted in posts to denote they're hyperlinks. Clicking on them should then create a search query with that word that highlights all instances across all posts on a global scope. That's it really. Basic hashtag functionality.

Optionally, the most relevant hashtags can be made visible at the top of the post in the same way topics currently are, without the need to add or remove them outside the richtext editor using context menu options. I think the problem here is that Humhub is a victim of its own versatility and excellence as an "all-in-one" social framework. Topics seem more relevant for forums or Facebook-style networks that have hierarchical sorting whereas hashtags are more for Twitter/IG style public-first networks.

Enhanced hashtags support would also include some way of quantifying the popularity of hashtags and showing that as a "popular topics" or "trending" widget on the dashboard. This by itself is interesting, but it would allow for suggested tags in the richtext editor when you begin typing `#` like with user mentions.

Ultimately, I don't think this issue is nearly as complex as it's been considered. I think the only problem is how unimportant it is.

**Comment by ArchBlood** [comment]
> I agree with @marc-farre. While topics may be a powerful feature, that doesn't ultimately matter if it's not used. Rebranding it as hashtags and changing its frontend appearance is the best way forward IMO. That's the easiest and fastest way to solve both problems.
> 
> Hashtags need no UI for adding them to a post since they're added directly in the richtext editor. But they need to be highlighted in posts to denote they're hyperlinks. Clicking on them should then create a search query with that word that highlights all instances across all posts on a global scope. That's it really. Basic hashtag functionality.
> 
> Optionally, the most relevant hashtags can be made visible at the top of the post in the same way topics currently are, without the need to add or remove them outside the richtext editor using context menu options. I think the problem here is that Humhub is a victim of its own versatility and excellence as an "all-in-one" social framework. Topics seem more relevant for forums or Facebook-style networks that have hierarchical sorting whereas hashtags are more for Twitter/IG style public-first networks.
> 
> Enhanced hashtags support would also include some way of quantifying the popularity of hashtags and showing that as a "popular topics" or "trending" widget on the dashboard. This by itself is interesting, but it would allow for suggested tags in the richtext editor when you begin typing `#` like with user mentions.
> 
> Ultimately, I don't think this issue is nearly as complex as it's been considered. I think the only problem is how unimportant it is.

Although I am incline to agree with this, the final deciding vote is for HumHub, if they're not wanting the functionality due to the existing topics or a possible future revised version that is their call to make important or unimportant, I will back out of the development process for the function in general and focus on other things such as improving my modules.

**Comment by luke-** [comment]
From a UX perspective, I would want to avoid such a solution in the Marketplace.
If so, it should integrate better with the current topic system. But here a core-side optimisation would make more sense.

**Comment by samuk** [comment]
It's been almost two years since we last talked about hashtags.

Is the 'vision' any closer? Or should I consider the third-party hashtags module if I want something working by the end of 2026?

**Comment by ArchBlood** [comment]
> It's been almost two years since we last talked about hashtags.
> 
> Is the 'vision' any closer? Or should I consider the third-party hashtags module if I want something working by the end of 2026?

The current version available is updated to 1.18 but isn't a part of HumHub's, Cuzy.app or Green Meteor's marketplace/store. You'll have to discuss that with TGSR OM from the HumHub community site.
